chore(developer): improve code coverage

srl295 opened the pull request:
- add a test for LdmlKeyboardCompiler.run()
- exclude generated abnf files from coverage

re: #13418 and https://github.com/keymanapp/keyman/pull/13498#issuecomment-2723252715

@keymanapp-test-bot skip

mcdurdin commented:
> * brings kmc-ldml to 90%+

Hurray! Please update build.sh to enforce the higher threshold:

```sh
builder_run_action test            builder_do_typescript_tests 90
```

or (as 90% is the default):

```sh
builder_run_action test            builder_do_typescript_tests
```

mcdurdin commented:
Failing build:

```
      ENOENT: no such file or directory, open 'basic-compiler-runB4PfUj/basic-run.kmx'
05:29:34       Error: ENOENT: no such file or directory, open 'basic-compiler-runB4PfUj/basic-run.kmx'
          at Object.openSync (node:fs:573:18)
          at readFileSync (node:fs:452:35)
          at Context.<anonymous> (file:///var/lib/TeamCity/work/30ffb35c1b6c4089/keyman/developer/src/kmc-ldml/test/compiler-e2e.tests.ts:59:25)
          at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

srl295 commented on developer/src/kmc-ldml/test/compiler-e2e.tests.ts:
```diff
@@ -44,6 +45,20 @@ describe('compiler-tests', function() {
+    const k = new LdmlKeyboardCompiler();
+    await k.init(compilerTestCallbacks, { ...compilerTestOptions, saveDebug: true, shouldAddCompilerVersion: false });
+
+    const tempdir = mkdtempSync('basic-compiler-run');
+    const outputFilename2 = join(tempdir, 'basic-run.kmx');
```
I'm trying to test run() which takes an output file

srl295 replied:
Maybe the 2nd parameter could be optional, where it returns a buffer instead…?

mcdurdin commented on developer/src/kmc-ldml/test/compiler-e2e.tests.ts:
```diff
@@ -44,6 +45,20 @@ describe('compiler-tests', function() {
+    const tempdir = mkdtempSync('basic-compiler-run');
+    const outputFilename2 = join(tempdir, 'basic-run.kmx');
+
+    await k.run(inputFilename, outputFilename);
```
Note that this does not write files out. It returns a `Promise<LdmlKeyboardCompilerResult>` which contains an `artifacts` object with `kmx?: Uint8Array` and `kvk?: Uint8Array` properties. They would not be written to disk unless you later call `LdmlKeyboardCompiler.write` with them (and even that use the callback to actually write to disk, so you can stub that out too if you need to. The compiler has no direct file system access.

srl295 replied:
yeah, i realizeded that. However, it doesn't match basic.kmx - it contains a .kvk file and one extra store! So, challenging to test.

srl295 commented:
@mcdurdin `run()` adds this little bit (plus the kvk file) … so, i can't compare the binary output against basic.kmx 

```js
    if(vkData) {
      kmx.keyboard.stores.push({
        dpName: '',
        dpString: keyboardId + '.kvk',
        dwSystemID: KMX.KMXFile.TSS_VISUALKEYBOARD
      });
    }
```

[here](https://github.com/keymanapp/keyman/blob/4e3379c69886d82f4e84f5b495913c9d2ee068d5/developer/src/kmc-ldml/src/compiler/compiler.ts#L155-L162)

mcdurdin commented:
> @mcdurdin `run()` inserts this little bit (plus the kvk file) … so, i can't compare the binary output against basic.kmx.

For ref, that was implemented in `run()` in #12402 to ensure that we get a visual keyboard for LDML keyboards.

Because basic.txt was being tested only with `compileKeyboard` and not `run`, that means there is divergence.

> This PR right now checks that `run()` generates _something_. Good for now?

I would like to fix this test.

Question: do we need to call both `run()` and the more internal `compile()` in this test? Seems like duplication? That is, perhaps this test becomes a whole lot shorter: given it is an e2e test, I think it's probably best that we compare the _final_ result of `run()`. `run()` is the public interface for kmc-ldml. While we are at it, we should cleanup the publics that are only exposed for unit tests on `LdmlKeyboardCompiler` to make them private, and expose them via a helper, like this:

https://github.com/keymanapp/keyman/blob/2a728b49f86dda02254bf8d0ffd9fdd6a3ad9ba4/developer/src/kmc-copy/src/KeymanProjectCopier.ts#L725-L729

srl295 commented:
> > @mcdurdin `run()` inserts this little bit (plus the kvk file) … so, i can't compare the binary output against basic.kmx.
> 
> For ref, that was implemented in `run()` in #12402 to ensure that we get a visual keyboard for LDML keyboards.
> 
> Because basic.txt was being tested only with `compileKeyboard` and not `run`, that means there is divergence.
> 
> > This PR right now checks that `run()` generates _something_. Good for now?
> 
> I would like to fix this test.

+1

> Question: do we need to call both `run()` and the more internal `compile()` in this test? Seems like duplication? 

run() does a bit more. so it's probably worth testing it also.

> That is, perhaps this test becomes a whole lot shorter: given it is an e2e test, I think it's probably best that we compare the _final_ result of `run()`. `run()` is the public interface for kmc-ldml. While we are at it, we should cleanup the publics that are only exposed for unit tests on `LdmlKeyboardCompiler` to make them private, and expose them via a helper, like this:
> 
> https://github.com/keymanapp/keyman/blob/2a728b49f86dda02254bf8d0ffd9fdd6a3ad9ba4/developer/src/kmc-copy/src/KeymanProjectCopier.ts#L725-L729

okay, i think you might mean things like compile() etc.


How about this, what about adding an additional fixture, just a generated .kmx file to stabilize the output of run()? (Rather than making a new basic.txt which includes kvk?)

mcdurdin commented:
> run() does a bit more. so it's probably worth testing it also.

Yeah, sorry if I wasn't clear. My thinking is we test only `run()` and not `compile()` etc in this test. That more closely matches how other consumers of the module will use it, so is appropriate IMHO.

> okay, i think you might mean things like compile() etc.
> 
> How about this, what about adding an additional fixture, just a generated .kmx file to stabilize the output of run()? (Rather than making a new basic.txt which includes kvk?)

If we test only `run()`, then we can just add the kvk data to the existing basic.txt, as this test is the only place where basic.txt is used.

srl295 commented:
- we don't have other tests of compile()
- we need the kmx+ output of compile() to test the serializer

- basic-txt.kmx and basic-xml.kmx diverged! Fixed that and added a test in the fixture generator.

mcdurdin commented on developer/src/kmc-ldml/build.sh:
```diff
@@ -75,14 +75,19 @@ function do_build_fixtures() {
   # Generate a binary file from basic.txt for comparison purposes
   node ../../../common/tools/hextobin/build/hextobin.js ./test/fixtures/basic.txt ./build/test/fixtures/basic-txt.kmx
+  if ! cmp ./build/test/fixtures/basic-xml.kmx ./build/test/fixtures/basic-txt.kmx; then
+    printf "${COLOR_RED}basic-xml.kmx and basic-txt.kmx are different!${COLOR_RESET}\n"
```
This should be a `builder_die` to fail here. It's a problem to fix

```suggestion
    builder_die "basic-xml.kmx and basic-txt.kmx are different!"
```

srl295 replied:
i went with this, and also added some hints if someone randomly gets this message in the future (mentioning basic.txt and basic.xml)

mcdurdin commented on developer/src/kmc-ldml/build.sh:
```diff
@@ -75,14 +75,19 @@ function do_build_fixtures() {
+  if ! cmp ./build/test/fixtures/basic-xml.kmx ./build/test/fixtures/basic-txt.kmx; then
+    printf "${COLOR_RED}basic-xml.kmx and basic-txt.kmx are different!${COLOR_RESET}\n"
+  else
+    printf "${COLOR_GREY}basic-xml.kmx and basic-txt.kmx are identical${COLOR_RESET}\n"
```
```suggestion
    builder_echo grey "basic-xml.kmx and basic-txt.kmx are identical"
```

Use `builder_echo` and `builder_die` rather than printf or echo directly

mcdurdin reviewed (approved):
LGTM

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 19.0.18-alpha](https://keyman.com/downloads/releases/alpha/19.0.18)
